server/src/main/java/org/opensearch/action/support/RetryableAction.java

```diff
@@ -142,18 +149,59 @@ public void onRejection(Exception e) {
 
     public void onFinished() {}
 
-    private class RetryingListener implements ActionListener<Response> {
+    private class CustomRetryingListener extends RetryingListener {
 
-        private static final int MAX_EXCEPTIONS = 4;
+        private Iterator<TimeValue> backoffDelayIterator;
+        private ArrayDeque<Exception> caughtExceptions;
+
+        private CustomRetryingListener(Iterator<TimeValue> backoffDelayIterator, ArrayDeque<Exception> caughtExceptions) {
+            super(caughtExceptions);
+            this.backoffDelayIterator = backoffDelayIterator;
+            this.caughtExceptions = caughtExceptions;
+        }
+
+        public RetryingListener getRetryingListenerForNextRetry() {
+            return this;
+        }
+
+        public long getRetryDelay() {
+            return backoffDelayIterator.next().millis();
+        }
+
+    }
+
+    private class DefaultRetryingListener extends RetryingListener {
```

Comment:
CustomRetryingListener is non-intuitive, rather can we do something like
1. Have overloaded constructor to take in `Iterator<TimeValue> backoffDelayIterator` in the existing `RetryingListener`
2. Move the logic to select the next RetryListener based on the overloaded values

Reply:
Thanks @Bukhtawar  for suggestion, Yes Initially I have implemented with this approach only. In that, based on null checks we will need decide the retry logic, which seems little complicated to understand from code and error prone as well. Below is sample snippet for it. 

```
        private ArrayDeque<Exception> caughtExceptions;
        private Iterator<TimeValue> backoffDelayIterator;
        private final long delayMillisBound;
        
        private RetryingListener(long delayMillisBound, ArrayDeque<Exception> caughtExceptions) {
            this.delayMillisBound = delayMillisBound;
            this.caughtExceptions = caughtExceptions;
        }

        private RetryingListener(Iterator<TimeValue> backoffDelayIterator, ArrayDeque<Exception> caughtExceptions) {
            this.backoffDelayIterator = backoffDelayIterator;
            this.caughtExceptions = caughtExceptions;
        }

        public RetryingListener getRetryingListenerForNextRetry() {
            if(backoffDelayIterator != null) {
                new RetryingListener(backoffDelayIterator);
            }
            final long nextDelayMillisBound = Math.min(delayMillisBound * 2, Integer.MAX_VALUE);
            final RetryingListener retryingListener = new DefaultRetryingListener(nextDelayMillisBound, caughtExceptions);
            return retryingListener;
        }

        public long getRetryDelay() {
            if(backoffDelayIterator != null){
                return backoffDelayIterator.next();
            }
            return delayMillisBound;
        }


```

So I decided to go with this approach and abstracted out the retry mechanism to sub class which looked cleaner way to me. Also those sub class are private as well so caller of `RetryableAction` is not need to worry about it, it will be handled internally only.

I am open to change it with above approach as well, but it doesn't seem cleaner way to me. Let me know your thoughts.

Reply:
should we have two constructors for `RetryableAction` : 

1.  taking `initialDelayMillis` :  creates DefaultRetryingListener 
2. takes `backoffPolicy` : creates CustomRetryingListener .

Reply:
@dhwanilpatel : The current RetryingListener also has a retry strategy which is exponential by the factor of 2 and is defined inline in Listener itself. Can we create a backpolicy for it as well, then there will be only one RetryingListener which takes backoff policy and would simplify the logic as well. For existing RetryableAction implementation will pick this new default back policy which was inline earlier.

Reply:
@shwetathareja , Will create the retry policy for existing logic and will use it. That will simplify it, as we need to pass policy only so.